README.md
# Obi-Wan Kenobot
This is a chatbot that mimics Jedi Master Obi-Wan Kenobi from Star Wars episodes 1-3.

![Obi-Wan Kenobi](https://www.tvinsider.com/wp-content/uploads/2021/12/BOBA_FETT_STAR_WARS_5-1014x570.jpg)

## Welcome!
Obi-Wan is the charismatic Jedi Master played by Ewan McGregor in the Star Wars prequel trilogy and features many iconic
lines such as "Oh no. I'm not brave enough for politics" and "Sith Lords are our specialty." This bot aims to decipher 
as best it can the user's intention when speaking to it, and respond with a quote that corresponds to that intent. The 
bot initially started as only a chatbot, but with the release of 2.0 the bot now has other features as well! As of right 
now, the bot is only available on Discord.

## Adding bot to your Discord server
This is the simplest part. Make sure you are an admin on whatever server you want to add it to and click this link
[here](https://discord.com/api/oauth2/authorize?client_id=956584753874743368&permissions=414464609344&scope=bot).

## How it works
The model uses the tensorflow machine learning library to create a text classification model. It then categorizes the
users input and sends out a random response for that category. To use the bot in your Discord server use the command: 
   
__ob!chat \<message\>__  

where __\<message\>__ is replaced with the message you want to send

## Other Features
* __ob!help__
    * Displays all commands and what they do
* __ob!talk__
  * Will join the channel the user is currently in and say a movie quote
* __ob!sjj \<number of players\>__
    * Starts a game of secret jar jar
    * __\<number of players\>__ is replaced with desired amount of players
        * Defaults to 5 which is the minimum
    * NOT FINISHED SO DOES NOTHING RIGHT NOW

## Sources
The data file was created myself using quotes found from the scripts found here: https://imsdb.com/ .   
Audio was taken from Youtube videos. The urls for those can be found in src/bot.py in the obitalk function.

## Authors
Matt Anikiej

CHANGELOG.md
# Version 2.3.1
* Changed command prefix to ob! from $
* Changed some commands to be easier to use (in README.md)
* Updated README.md
* added docstring and option to select number of players in secret jarjar

# Version 2.3.0
* Changed command prefix to $ from /
* added help command
* added skeleton of secret jar jar

# Version 2.2.0
* chatbot 2.0!!
* Many improvements have been made to the model that obi wan uses
* trained until loss didn't improve for 5 epochs
* does not need to train every time the bot starts running
    * SIGNIFICANTLY reduces computation strain
* new test function added to Model

# Version 2.1.0
* fixed apostrophe bug (it just disappeared)
* changed /obiwan to /obichat since it's not the only feature of the bot anymore
* removed bug log from repo since there were none anymore

# Version 2.0.1
* fixed README.md

# Version 2.0.0
* Brand new feature!!
* Obi-Wan Kenobot can now say an iconic quote in the voice chat!
* Simply connect to a voice chat and then type /obitalk and Kenobot will join to say a line!
* also minor cahnges to make code more readable
* requirements.txt updated
* README.md updated

# Version 1.0.1
* bot has been deployed on AWS instead of Heroku
* removed Procfile

# Version 1.0.0
* bot has been deployed!!
* Procfile added for deployment
* requirements.txt finalized
* README.md updated

# Version 0.2.1
* bot now responds to user input

# Version 0.2.0
* app is now a discord bot
    * not deployed yet
* model folder is now a package
* data folder is moved to under source folder
* Model is now built by specifying data path
* requirements.txt added
* bot responds with 'hello' when /obiwan command is used

# Version 0.1.0
* model.py is finished and fully functional
  * model can be:
    * printed
    * trained
    * saved
    * loaded

# Version 0.0.3
* model.py can be trained and printed as a string
* Model.py renamed to model.py
* data folder moved to model folder

# Version 0.0.2
* Model.py will create the model and show the summary

# Version 0.0.1
* The class containing the model was created

src/bot.py
import discord
from discord.ext import commands

import os
from time import sleep
import random

from model.model import Model
from SecretJarJar.secret_jarjar import SecretJarJar
from YTDLSource import YTDLSource

# sets up client and command prefix
client = commands.Bot(command_prefix='ob!', help_command=None)
token = os.getenv("OBI_WAN_KENOBOT_TOKEN")

# load pretrained model in
model = Model(data_path='data/obiwanintents.json')
model.load_model('model/saved_models/obiwankenobot')


@client.event
async def on_ready():
    """
    controls bot behavior on start up
    """
    await client.change_presence(status=discord.Status.online, activity=discord.Game("Podracing"))


# command ob!help
@client.command()
async def help(ctx):
    """
    Use case: ob!help
    Displays all commands
    :param ctx: context of the client
    """
    help_msg = """
    Obi-Wan Kenobot Commands (prefix=ob!):
    **chat <message>** -  use to send a message to obiwan and he will respond in chat
    **talk** - use while in a voice channel to invite obiwan and have him say a line
    **sjj** - use to play Secret JarJar NOT FINISHED
    """
    await ctx.send(help_msg)


# command ob!chat
@client.command()
async def chat(ctx, *message):
    """
    Use case: ob!chat <message>
    Used to talk to the chat bot by sending a message, and then the bot replies.
    If no message is sent then the bot responds with 'Hello there!'
    :param ctx: context of the client
    :param message: message user sends the bot
    """
    if len(message) == 0:
        await ctx.send("Hello there! Send me a message after the command: /obichat message")
    else:
        # message is a tuple of variable length where each word is an index and must be joined into one string
        message_str = ' '.join(message)
        response = model.chat(message_str)
        await ctx.send(response)


# command ob!talk
@client.command(pass_context=True)
async def talk(ctx):
    """
    Use case: ob!talk
    Used for obiwan to talk in the voice channel the user is currently connected to
    :param ctx: context of the call
    """

    # make sure user is connected to a voice chat
    if not ctx.message.author.voice:
        await ctx.send("{} is not connected to a voice channel".format(ctx.message.author.name))
        return
    else:
        channel = ctx.message.author.voice.channel
        await channel.connect()

    # set of urls for the bot to choose from
    urls = [
            'https://www.youtube.com/watch?v=iIQUrrbNW_Q&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=58JwbQCnpmA&list=PLlrQRD4Rfv_CM6byFKlMYeBWAs9dWDm_7&index=2&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=Xzbyp8rjGLU&list=PLlrQRD4Rfv_CM6byFKlMYeBWAs9dWDm_7&index=10&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=Pqn3zMghf4A&list=PLlrQRD4Rfv_CM6byFKlMYeBWAs9dWDm_7&index=21&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=aNmZ9U8lE6g&list=PLlrQRD4Rfv_CM6byFKlMYeBWAs9dWDm_7&index=36&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=41hvkx77nAU&list=PLlrQRD4Rfv_CVrEGTWM3r9EZa3jYUXs0N&index=5&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=LH2-Ha6PgpA&list=PLlrQRD4Rfv_CVrEGTWM3r9EZa3jYUXs0N&index=35&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=v-iDFxFC_9E&list=PLlrQRD4Rfv_CVrEGTWM3r9EZa3jYUXs0N&index=36&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=frszEJb0aOo&list=PLlrQRD4Rfv_CNLUjhLBJg_pUfq207fPyD&index=7&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=BOvcSLpOp4o&list=PLlrQRD4Rfv_CNLUjhLBJg_pUfq207fPyD&index=8&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=37Q0fx5r-H8&list=PLlrQRD4Rfv_CNLUjhLBJg_pUfq207fPyD&index=15&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=jBOMlWl7fFk&list=PLlrQRD4Rfv_CNLUjhLBJg_pUfq207fPyD&index=17&ab_channel=QuoteTheGuy',
            'https://www.youtube.com/watch?v=J0BciHfsU7k&list=PLlrQRD4Rfv_CNLUjhLBJg_pUfq207fPyD&index=22&ab_channel=QuoteTheGuy'
            ]

    # select a random url
    i = random.randint(0, len(urls)-1)
    url = urls[i]

    voice_channel = ctx.message.guild.voice_client

    player = await YTDLSource.from_url(url, loop=client.loop)

    voice_channel.play(player, after=lambda e: print('Player error: %s' % e) if e else None)

    # don't have bot disconnect before it finishes
    while voice_channel.is_playing():
        sleep(1)

    if voice_channel.is_connected():
        await voice_channel.disconnect()


# command ob!sjj
@client.command()
async def sjj(ctx, num_players=5):
    """
    Use case: ob!sjj <num_players>
    Used to start a game of secret jar jar.
    The game begins when all players reacted to the message to join.
    :param ctx: context of the command
    :param num_players: number of players to join. Must be >= 5
    """
    # minimum number of players is 5
    if num_players < 5:
        await ctx.send('How did this happen, we\'re smarter than this! The number of players must be greater than 5')
        return
    game = SecretJarJar([num_players])

# run bot
client.run(token)


src/YTDLSource.py
import discord
import youtube_dl

youtube_dl.utils.bug_reports_message = lambda: ''


ytdl_format_options = {
    'format': 'bestaudio/best',
    'outtmpl': 'audio/%(extractor)s-%(id)s-%(title)s.%(ext)s',
    'restrictfilenames': True,
    'noplaylist': True,
    'nocheckcertificate': True,
    'ignoreerrors': False,
    'logtostderr': False,
    'quiet': True,
    'no_warnings': True,
    'default_search': 'auto',
    'source_address': '0.0.0.0' # bind to ipv4 since ipv6 addresses cause issues sometimes
}

ffmpeg_options = {
    'options': '-vn'
}

ytdl = youtube_dl.YoutubeDL(ytdl_format_options)


class YTDLSource(discord.PCMVolumeTransformer):
    def __init__(self, source, *, data, volume=0.5):
        super().__init__(source, volume)

        self.data = data

        self.title = data.get('title')
        self.url = data.get('url')

    @classmethod
    async def from_url(cls, url, *, loop=None, stream=False):
        loop = loop or asyncio.get_event_loop()
        data = await loop.run_in_executor(None, lambda: ytdl.extract_info(url, download=not stream))

        if 'entries' in data:
            # take first item from a playlist
            data = data['entries'][0]

        filename = data['url'] if stream else ytdl.prepare_filename(data)
        return cls(discord.FFmpegPCMAudio(filename, **ffmpeg_options), data=data)

src/model/Model.py
import json
import re

import pandas as pd
import numpy as np

import tensorflow as tf
from tensorflow import keras
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences


class Model:
    """
    This class houses the text classification model for the chatbot
    """

    #
    # CONSTRUCTORS
    #
    def __init__(self):
        """
        Constructor for the model
        """
        self.data_ = None  # data that the model uses to train on
        self.model_ = None  # model that takes in user input and outputs intent
        self.label_map_ = None  # maps output to label
        self.responses_ = None  # maps label to list of responses
        self.tokenizer_ = None  # tokenizer used to change strings to vectors with numbers

        self.__load_data()
        self.__create_model()

    #
    # PRIVATE FUNCTIONS
    #
    def __standardize_text(self, string):
        """
        Makes string lower case and removes punctuation
        :param string: string to be converted
        :return: converted string
        """
        return re.sub(r'[^\w\s]', '', string).lower()

    def __load_data(self):
        """
        loads in the data from the intents file,
        and returns a pandas data frame and sets the data and label_map members
        """
        with open('datasets/obiwanintents.json') as file:
            data = json.load(file)

        # parse through data
        responses = {}
        inputs = []
        labels = []
        label_map = {}

        count = 0
        for intent in data['intents']:
            responses[intent['tag']] = intent['responses']
            label_map[intent['tag']] = count
            count += 1
            for message in intent['inputs']:
                inputs.append(message)
                labels.append(intent['tag'])

        # store responses
        self.responses_ = responses

        # converting to dataframe
        data = pd.DataFrame({"inputs": inputs,
                             "labels": labels})

        # shuffle the data
        data = data.sample(frac=1).reset_index(drop=True)

        data['labels'].replace(label_map, inplace=True)

        # standardize all text to be lowercase and no punctuation
        for i in range(len(data)):
            data['inputs'].iloc[i] = self.__standardize_text(data['inputs'].iloc[i])

        # reverse the label map so each number is the key
        # this needs to be reversed since output of model is the number
        self.label_map_ = dict((v, k) for k, v in label_map.items())
        self.data_ = data

    def __tokenize_initial_data(self):
        """
        Uses the training data to fit a tokenizer
        :return: tokenized data
        """
        # allow up to 10000 unique words
        self.tokenizer = Tokenizer(num_words=10000)
        self.tokenizer.fit_on_texts(self.data['inputs'])
        X = self.tokenizer.texts_to_sequences(self.data['inputs'])
        X = pad_sequences(X)
        return X

    def __create_model(self, embedding_dim=64):
        X = self.tokenize_initial_data()

        # get input shape from teh tokenized data
        # shape can change with changes in the data so this is the best way for future proofing the model
        input_shape = X.shape[1]

        inputs = keras.Input(input_shape, name='input')

        # embedding layer to create a feature map of the initial vector
        lyr = keras.layers.Embedding(input_dim=len(self.tokenizer_.word_index) + 1, output_dim=embedding_dim,
                                     name='embedding')(inputs)
        # LSTM layer is the best way to understand NLP due to its memory components, so it can understand sequences
        lyr = keras.layers.LSTM(embedding_dim, return_sequences=True, name='lstm')(lyr)

        # flatten to 1D so it's easier on the dense layer
        lyr = keras.layers.Flatten(name='flatten')(lyr)

        # use softmax activation since all the labels are mutually exclusive
        outputs = keras.layers.Dense(13, activation='softmax', name='output')(lyr)

        self.model_ = keras.Model(inputs=inputs, outputs=outputs, name='ObiWan_Kenobot')
        self.model_.summary()

model = Model()


src/SecretJarJar/player.py
class Player:
    """
    Class that describes a player in secret jarjar
    """
    def __init__(self, name):
        """
        Constructor for players for secret jarjar
        :param name: name of the player
        """
        self.name = name
        self.role = ''
        self.team = ''
        self.master = False
        self.apprentice = False

    def get_name(self):
        """
        Gets the name of the player
        :return:
        """
        return self.name

    def set_name(self, name):
        """
        Sets the name of the player
        :param name: name to set the player's name to
        """
        self.name = name

    def get_role(self):
        """
        Gets the role of the player
        :return:
        """
        return self.role

    def set_role(self, role):
        """
        Sets the role of the player
        :param role: role to set the player's role to
        """
        self.role = role

    def get_team(self):
        """
        Gets the team of the player
        :return:
        """
        return self.team

    def set_team(self, team):
        """
        Sets the team of the player
        :param team: team to set the player's team to
        """
        self.team = team

    def set_master(self, master):
        """
        Sets the master of the player
        :param master: master to set the player's master to
        """
        self.master = master

    def set_apprentice(self, apprentice):
        """
        Sets the apprentice of the player
        :param apprentice: apprentice to set the player's apprentice to
        """
        self.apprentice = apprentice


src/SecretJarJar/game_board.py


class GameBoard:
    """
    Class that describes the game board
    """
    def __init__(self, num_players):
        """
        Constructor for the game board
        :param num_players: how many players are in the game
        """
        self.num_players = num_players
        # initialize board to start
        self.board = self.create_board()

    def __str__(self):
        """
        String representation of GameBoard
        :return: string of GameBoard
        """
        space_map = {-1: 'X', 0: ' ', 1: 'ACTION'}
        gb = f"""
        ```
        ---------------------------
        | {'JEDI':^10} | {'SITH':^10} |
        ---------------------------
        | {space_map[self.board[0][0]]:^10} | {space_map[self.board[1][0]]:^10} |
        | {space_map[self.board[0][1]]:^10} | {space_map[self.board[1][1]]:^10} |
        | {space_map[self.board[0][2]]:^10} | {space_map[self.board[1][2]]:^10} |
        | {space_map[self.board[0][3]]:^10} | {space_map[self.board[1][3]]:^10} |
        | {space_map[self.board[0][4]]:^10} | {space_map[self.board[1][4]]:^10} |
        ---------------------------
        | {space_map[self.board[0][5]]:^10} | {space_map[self.board[1][5]]:^10} |
        ---------------------------
        ```
        """
        return gb

    def create_board(self):
        """
        Makes the game board
        with 2 actions per team
        :param num_players: number of players in the game
        :return: empty game board
        """
        jedi_board = [0, 0, 0, 0, 0, 0]
        sith_board = [0, 0, 0, 0, 0, 0]
        if self.num_players < 6:
            # 2 actions per board
            jedi_board[3] = 1
            jedi_board[4] = 1

            sith_board[3] = 1
            sith_board[4] = 1

        elif self.num_players < 9:
            # 3 actions per board
            jedi_board[2] = 1
            jedi_board[3] = 1
            jedi_board[4] = 1

            sith_board[2] = 1
            sith_board[3] = 1
            sith_board[4] = 1

        else:
            # actions on every policy
            for i in range(5):
                jedi_board[i] = 1
                sith_board[i] = 1

        board = [jedi_board, sith_board]
        return board

    def add_policy(self, team):
        """
        Adds policy to the board
        :param team: team to add policy to
        :return: True if there is an action, false otherwise
        """
        t = 0
        if team.lower() == 'sith':
            t = 1

        action = False
        # loops until first empty spot is found
        for i in range(len(self.board[t])):
            if self.board[t][i] != -1:
                if self.board[t][i] == 1:
                    action = True
                self.board[t][i] = -1
                break

        return action


src/model/model.py
import json
import re
import random

import pandas as pd

import tensorflow as tf
from tensorflow import keras
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences


class Model:
    """
    This class houses the text classification model for the chatbot
    """

    #
    # CONSTRUCTORS
    #
    def __init__(self, data_path='../data/obiwanintents.json'):
        """
        Constructor for the model
        """
        self.data_path_ = data_path  # path where data is located
        self.data_ = None  # data that the model uses to train on
        self.label_map_ = None  # maps output to label
        self.responses_ = None  # maps label to list of responses

        self.tokenizer_ = None  # tokenizer used to change a string to a vector with numbers
        self.input_shape_ = None  # shape of the input, needed for padding user inputs

        self.model_ = None  # model that takes in user input and outputs intent

        self.__load_data(data_path)  # initializes: data_ label_map_ responses_
        self.__tokenize_initial_data()  # initializes: tokenizer: input_shape_
        self.__create_model()  # initializes: model_

    def __str__(self):
        """
        String conversion of model
        """
        self.model_.summary()
        return ''

    #
    # PRIVATE FUNCTIONS
    #
    def __standardize_text(self, string):
        """
        Makes string lower case and removes punctuation
        :param string: string to be converted
        :return: converted string
        """
        return re.sub(r'[^\w\s]', '', string).lower()

    def __load_data(self, name):
        """
        loads in the data from the intents file,
        and returns a pandas data frame and sets the data and label_map members
        :param name: name of the file to load
        """
        with open(name) as file:
            data = json.load(file)

        # parse through data
        responses = {}
        inputs = []
        labels = []
        label_map = {}

        count = 0
        for intent in data['intents']:
            responses[intent['tag']] = intent['responses']
            label_map[intent['tag']] = count
            count += 1
            for message in intent['inputs']:
                inputs.append(message)
                labels.append(intent['tag'])

        # store responses
        self.responses_ = responses

        # converting to dataframe
        data = pd.DataFrame({"inputs": inputs,
                             "labels": labels})

        data['labels'].replace(label_map, inplace=True)

        # standardize all text to be lowercase and no punctuation
        for i in range(len(data)):
            data['inputs'].iloc[i] = self.__standardize_text(data['inputs'].iloc[i])

        # reverse the label map so each number is the key
        # this needs to be reversed since output of model is the number
        self.label_map_ = dict((v, k) for k, v in label_map.items())
        self.data_ = data

    def __tokenize_initial_data(self):
        """
        Uses the training data to fit a tokenizer
        :return: tokenized data
        """
        # allow up to 10000 unique words
        self.tokenizer_ = Tokenizer(num_words=10000)
        self.tokenizer_.fit_on_texts(self.data_['inputs'])
        X = self.tokenizer_.texts_to_sequences(self.data_['inputs'])
        X = pad_sequences(X)

        # get input shape from the tokenized data
        # shape can change with changes in the data so this is the best way for future proofing the model
        self.input_shape_ = X.shape[1]

        return X

    def __create_model(self, embedding_dim=64):
        """
        Creates the model
        :param embedding_dim: number of dimensions for embedding layer
        """

        # MODEL ARCHITECTURE:
        #
        # embedding layer to create a feature map of the initial vector
        #
        # LSTM layer is the best way to understand NLP,
        # due to its memory components that allow it to understand sequences
        #
        # flatten to 1D so that it's easier on the dense layer
        #
        # use softmax activation since all the labels are mutually exclusive
        inputs = keras.Input(self.input_shape_, name='input')

        lyr = keras.layers.Embedding(input_dim=len(self.tokenizer_.word_index) + 1, output_dim=embedding_dim,
                                     name='embedding')(inputs)
        lyr = keras.layers.LSTM(embedding_dim, name='lstm')(lyr)
        lyr = keras.layers.Flatten(name='flatten')(lyr)
        outputs = keras.layers.Dense(13, activation='softmax', name='output')(lyr)

        self.model_ = keras.Model(inputs=inputs, outputs=outputs, name='ObiWan_Kenobot')

        # compiling the model
        # sparse categorical cross entropy loss since each category is mutually exclusive
        self.model_.compile(loss="sparse_categorical_crossentropy",
                            optimizer='adam', metrics=['accuracy'])

    #
    # PUBLIC FUNCTIONS
    #
    def summary(self):
        """
        returns the summary of the model
        :return: summary of the model
        """
        return self.model_.summary()

    def train(self):
        """
        Trains the model on the given data
        :return: history of losses and metrics
        """
        X = self.__tokenize_initial_data()
        return self.model_.fit(X, self.data_['labels'],
                               epochs=10000,
                               callbacks=[tf.keras.callbacks.EarlyStopping(patience=5, monitor='loss')])

    def save_model(self, name="model"):
        """
        saves the weights to a .ckpt file with the given name to the directory saved_models
        :param name: name of the file
        """
        self.model_.save_weights('saved_models/'+name)

    def load_model(self, path='saved_models/model'):
        """
        loads the weights from the file with the given name
        :param path: path to the file to load in
        """
        self.model_.load_weights(path)
        self.model_.compile(loss="sparse_categorical_crossentropy",
                            optimizer='adam', metrics=['accuracy'])

    def chat(self, user_input):
        """
        Accepts a string input from the user and returns a response that corresponds to the predicted intent
        :param user_input: string that the user sends the bot
        :return: string that the bot responds with
        """
        # vectorize input to be able to use in the model
        user_input = self.__standardize_text(user_input)
        vec = self.tokenizer_.texts_to_sequences([user_input])
        vec = [item for i in vec for item in i]  # needs to be flattened for padding
        vec = pad_sequences([vec], self.input_shape_)

        # get label
        label = self.model_.predict(vec)
        label = label.argmax()
        tag = self.label_map_[label]

        # return a random corresponding response
        responses = self.responses_[tag]
        i = random.randint(0, len(responses)-1)
        return responses[i]

    def test(self):
        """
        Used to test the accuracy of the model
        """
        greeting = ['how are you', 'greeting']
        goodbye = ['See you later.', 'goodbye']
        thanks = ['Thank you so much!', 'thanks']
        tasks = ['What can you do?', 'tasks']
        alive = ['Are you even alive at all?', 'alive']
        hlp = ['PLease I need your help!', 'help']
        mission = ['Can you tell me a little bit about your mission?', 'mission']
        jedi = ['are you a jedi?', 'jedi']
        sith = ['Who is the most evil sith in the galaxy?', 'sith']
        bounty = ['Do you know of Jango Fett?', 'bounty hunter']
        funny = ['I know you know some good jokes', 'funny']
        stories = ['Cmon, tell me a story!', 'stories']
        threat = ['I will fight you', 'threat']

        messages = [greeting, goodbye, thanks, tasks, alive, hlp, mission, jedi, sith, bounty, funny, stories, threat]

        correct = 0
        for msg in messages:
            msg[0] = self.__standardize_text(msg[0])
            X = self.tokenizer_.texts_to_sequences([msg[0]])
            X = [item for i in X for item in i]
            X = pad_sequences([X], self.input_shape_)
            pred = self.model_.predict(X)
            pred = pred.argmax()
            print('Pred: {} Actual: {}'.format(self.label_map_[pred], msg[1]))
            if self.label_map_[pred] == msg[1]:
                correct += 1
            print()

        accuracy = correct / len(messages)
        print('Accuracy = {:.2f}'.format(accuracy))
